packages/medusa/src/api/routes/store/returns/create-return.js

```diff
@@ -0,0 +1,188 @@
+
+    let inProgress = true
+    let err = false
+
+    while (inProgress) {
+      switch (idempotencyKey.recovery_point) {
+        case "started": {
+          const { key, error } = await idempotencyKeyService.workStage(
+            idempotencyKey.idempotency_key,
+            async manager => {
+              const order = await orderService
+                .withTransaction(manager)
+                .retrieve(value.order_id, {
+                  select: ["refunded_total", "total"],
+                  relations: ["items"],
+                })
+
+              const returnObj = {
+                order_id: value.order_id,
+                idempotency_key: idempotencyKey.idempotency_key,
+                items: value.items,
+              }
+
+              if (value.return_shipping) {
+                returnObj.shipping_method = value.return_shipping
+              }
+
+              if (typeof value.refund !== "undefined" && value.refund < 0) {
+                returnObj.refund_amount = 0
+              } else {
+                if (value.refund) {
+                  returnObj.refund_amount = value.refund
+                }
+              }
+
+              const createdReturn = await returnService
+                .withTransaction(manager)
+                .create(returnObj, order)
+
+              if (value.return_shipping) {
+                await returnService
+                  .withTransaction(manager)
+                  .fulfill(createdReturn.id)
+              }
+
+              await eventBus
+                .withTransaction(manager)
+                .emit("order.return_requested", {
+                  id: value.order_id,
+                  return_id: createdReturn.id,
+                })
+
+              return {
+                recovery_point: "return_requested",
+              }
+            }
+          )
+
+          if (error) {
+            inProgress = false
+            err = error
+          } else {
+            idempotencyKey = key
+          }
+          break
+        }
+
+        case "return_requested": {
+          const { key, error } = await idempotencyKeyService.workStage(
+            idempotencyKey.idempotency_key,
+            async manager => {
+              let order = await orderService
+                .withTransaction(manager)
+                .retrieve(value.order_id, { relations: ["returns"] })
+
+              let ret = await returnService.withTransaction(manager).list({
+                idempotency_key: idempotencyKey.idempotency_key,
+              })
+              if (!ret.length) {
+                throw new MedusaError(
+                  MedusaError.Types.INVALID_DATA,
+                  `Return not found`
+                )
+              }
+              ret = ret[0]
+
+              /**
+               * If we are ready to receive immediately, we find the newly created return
+               * and register it as received.
+               */
+              if (value.receive_now) {
```

Comment:
`receive_now` is missing in validation schema.

Reply:
Would be good to also add an integration test here.

Reply:
Have removed - same idea: it's a storefront endpoint so the customer shouldn't be able to decide if their goods are already received.